app/scripts/services/globalPromotionService.ts

```diff
@@ -6,30 +6,85 @@ import { GlobalPromotion } from 'globalPromotion';
 const mockUrl = 'http://localhost:9001';
 
 export class GlobalPromotionService {
+  private globalPromosCache: {
+    [key: string]: GlobalPromotion[];
+  } = {};
+
   /* @ngInject */
   constructor(
     private $rootScope: $RootScope,
     private $http: $Http,
     private modalMessage: ModalMessage,
   ) {}
 
-  loadPromoCodes(): IPromise<GlobalPromotion[]> {
+  private getCacheKey(
+    ministryId?: string,
+    ministryActivityId?: string,
+  ): string {
+    // API ignores ministryActivityId if ministryId is not provided
+    // So normalize the cache key to reflect actual API behavior
+    if (!ministryId) {
+      return 'all:all';
+    }
+    return `${ministryId}:${ministryActivityId || 'all'}`;
+  }
+
+  loadPromoCodes(
+    ministryId?: string,
+    ministryActivityId?: string,
+  ): IPromise<GlobalPromotion[]> {
+    const cacheKey = this.getCacheKey(ministryId, ministryActivityId);
+
+    if (this.globalPromosCache[cacheKey]) {
+      return Promise.resolve(this.globalPromosCache[cacheKey]);
```

Comment:
If we return a regular promise, it will message up AngularJS' state update checking.

```suggestion
      return $q.resolve(this.globalPromosCache[cacheKey]);
```